File: crates/ferrogate-observability/src/backend_test.rs
// Token4AI Cloud Attribution
// Developed by the commercial cloud service company represented by https://token4ai.cloud.
// Author: jamesduan (X: https://x.com/JamesDuanL)
// Created: 2026-07-26
// description: Token4AI Cloud, FerroGate AI Gateway, Rust API Gateway, agent-native AI traffic infrastructure.

use super::*;
use crate::otlp::OtlpAttribute;

fn snapshot() -> GatewayMetricsSnapshot {
    GatewayMetricsSnapshot {
        service_name: "ferrogate".to_string(),
        ..Default::default()
    }
}

fn span() -> OtlpSpanRecord {
    OtlpSpanRecord {
        trace_id: "0af7651916cd43dd8448eb211c80319c".to_string(),
        span_id: "b7ad6b7169203331".to_string(),
        parent_span_id: None,
        name: "ferrogate.gateway.request".to_string(),
        start_time_unix_nano: 1,
        end_time_unix_nano: 2,
        attributes: vec![OtlpAttribute::new("tenant", "acme")],
    }
}

fn log() -> OtlpLogRecord {
    OtlpLogRecord {
        trace_id: None,
        span_id: None,
        severity_text: "INFO".to_string(),
        body: "request".to_string(),
        time_unix_nano: 1,
        attributes: Vec::new(),
    }
}

#[test]
fn otlp_backend_builds_a_request_per_signal() {
    let backend = OtlpBackend::new("http://collector:4318");

    let metrics = backend.metrics_request(&snapshot()).unwrap().unwrap();
    assert_eq!(metrics.url, "http://collector:4318/v1/metrics");

    let traces = backend
        .traces_request("ferrogate", &[span()])
        .unwrap()
        .unwrap();
    assert_eq!(traces.url, "http://collector:4318/v1/traces");

    let logs = backend
        .logs_request("ferrogate", &[log()])
        .unwrap()
        .unwrap();
    assert_eq!(logs.url, "http://collector:4318/v1/logs");
}

#[test]
fn otlp_backend_skips_empty_batches() {
    let backend = OtlpBackend::new("http://collector:4318");

    assert!(backend.traces_request("ferrogate", &[]).unwrap().is_none());
    assert!(backend.logs_request("ferrogate", &[]).unwrap().is_none());
}

#[test]
fn otlp_backend_skips_signals_it_does_not_carry() {
    let backend =
        OtlpBackend::new("http://collector:4318").with_signals(vec![ObservabilitySignal::Metric]);

    assert!(backend.supports(ObservabilitySignal::Metric));
    assert!(!backend.supports(ObservabilitySignal::Trace));
    assert!(backend.metrics_request(&snapshot()).unwrap().is_some());
    assert!(backend
        .traces_request("ferrogate", &[span()])
        .unwrap()
        .is_none());
    assert!(backend
        .logs_request("ferrogate", &[log()])
        .unwrap()
        .is_none());
}

#[test]
fn otlp_backend_validate_rejects_a_scheme_less_endpoint() {
    let error = OtlpBackend::new("collector:4318").validate().unwrap_err();
    assert!(matches!(
        error,
        ObservabilityConfigError::InvalidEndpoint { .. }
    ));
}

#[test]
fn otlp_backend_validate_rejects_an_empty_endpoint() {
    let error = OtlpBackend::new("   ").validate().unwrap_err();
    assert!(matches!(
        error,
        ObservabilityConfigError::MissingEndpoint { .. }
    ));
}

#[test]
fn otlp_backend_carries_no_credential_headers() {
    let backend = OtlpBackend::new("http://collector:4318");
    let metrics = backend.metrics_request(&snapshot()).unwrap().unwrap();
    assert!(metrics.headers.is_empty());
}

#[test]
fn backends_are_usable_as_trait_objects() {
    // The point of the trait: the export loop holds one of these and never
    // matches on a destination enum.
    let backends: Vec<Box<dyn TelemetryBackend>> = vec![
        Box::new(OtlpBackend::new("http://collector:4318")),
        Box::new(crate::cloudflare::CloudflareBackend::new(
            "https://collector.example.workers.dev",
            "token",
        )),
    ];

    let names = backends
        .iter()
        .map(|backend| backend.name())
        .collect::<Vec<_>>();
    assert_eq!(names, vec!["otlp", "cloudflare"]);
    for backend in &backends {
        assert!(backend.validate().is_ok());
        assert!(backend.metrics_request(&snapshot()).unwrap().is_some());
    }
}

File: packages/observability/test/backend.test.ts
import { describe, expect, test } from "vitest";
import {
  CloudflareBackend,
  defaultGatewayMetricsSnapshot,
  ObservabilitySignal,
  OtlpBackend,
  otlpAttribute,
  type GatewayMetricsSnapshot,
  type OtlpLogRecord,
  type OtlpSpanRecord,
  type TelemetryBackend,
} from "../src/index.js";

function snapshot(): GatewayMetricsSnapshot {
  return { ...defaultGatewayMetricsSnapshot(), serviceName: "ferrogate" };
}

function span(): OtlpSpanRecord {
  return {
    traceId: "0af7651916cd43dd8448eb211c80319c",
    spanId: "b7ad6b7169203331",
    name: "ferrogate.gateway.request",
    startTimeUnixNano: 1,
    endTimeUnixNano: 2,
    attributes: [otlpAttribute("tenant", "acme")],
  };
}

function log(): OtlpLogRecord {
  return {
    severityText: "INFO",
    body: "request",
    timeUnixNano: 1,
    attributes: [],
  };
}

describe("OtlpBackend", () => {
  test("builds a request per signal", () => {
    const backend = new OtlpBackend("http://collector:4318");
    expect(backend.metricsRequest(snapshot())?.url).toBe(
      "http://collector:4318/v1/metrics",
    );
    expect(backend.tracesRequest("ferrogate", [span()])?.url).toBe(
      "http://collector:4318/v1/traces",
    );
    expect(backend.logsRequest("ferrogate", [log()])?.url).toBe(
      "http://collector:4318/v1/logs",
    );
  });

  test("skips empty batches", () => {
    const backend = new OtlpBackend("http://collector:4318");
    expect(backend.tracesRequest("ferrogate", [])).toBeNull();
    expect(backend.logsRequest("ferrogate", [])).toBeNull();
  });

  test("skips signals it does not carry", () => {
    const backend = new OtlpBackend("http://collector:4318").withSignals([
      ObservabilitySignal.Metric,
    ]);
    expect(backend.supports(ObservabilitySignal.Metric)).toBe(true);
    expect(backend.supports(ObservabilitySignal.Trace)).toBe(false);
    expect(backend.metricsRequest(snapshot())).not.toBeNull();
    expect(backend.tracesRequest("ferrogate", [span()])).toBeNull();
    expect(backend.logsRequest("ferrogate", [log()])).toBeNull();
  });

  test("validate rejects a scheme-less endpoint", () => {
    expect(new OtlpBackend("collector:4318").validate()?.errorKind).toBe(
      "InvalidEndpoint",
    );
  });

  test("validate rejects an empty endpoint", () => {
    expect(new OtlpBackend("   ").validate()?.errorKind).toBe("MissingEndpoint");
  });

  test("carries no credential headers", () => {
    const metrics = new OtlpBackend("http://collector:4318").metricsRequest(
      snapshot(),
    );
    expect(metrics?.headers).toEqual([]);
  });
});

describe("TelemetryBackend as a common contract", () => {
  test("backends are usable through the shared interface", () => {
    const backends: TelemetryBackend[] = [
      new OtlpBackend("http://collector:4318"),
      new CloudflareBackend(
        "https://collector.example.workers.dev",
        "token",
      ),
    ];
    expect(backends.map((b) => b.name())).toEqual(["otlp", "cloudflare"]);
    for (const backend of backends) {
      expect(backend.validate()).toBeNull();
      expect(backend.metricsRequest(snapshot())).not.toBeNull();
    }
  });
});
